### PR TITLE
Issue 24 add json metadata commit body

### LINKED ISSUE #24: Use a `json` commit body to include queue metadata
As we were discussing [here](https://github.com/josecelano/library-consumer/issues/4), it would be convenient to allow the queue to add metadata in the commit body. I think this was the final version:

### Commit body

NewJobMessage:

```
{
  "name": "create-job",
  "metadata": {
     "version": 1,
  }
  "payload": "{\"previous_ref\": \"...\",\n     \"current_ref\": \"...\",\n     \"submodule\": \"...\",\n     \"submoudle_url\": \"...\",\n}"
}
```

StartJobMessage:

```
{
  "name": "begin-job",
  "metadata": {
     "version": 1,
     "job_number": 1,
     "job_commit": "0fdf0db843a2010702b7c8d2e446966e9302f40e",
  },
  "payload": ""
}
```

FinishJobMessage:

```
{
  "name": "finish-job",
  "metadata": {
     "version": 1,
     "job_number": 1,
     "job_commit": "0fdf0db843a2010702b7c8d2e446966e9302f40e",
  },
  "payload": ""
}
```

@da2ce7 @yeraydavidrodriguez please review this.

### PR DESCRIPTION
**Pull request by yeraydavidrodriguez**
Resolves #24 

This will add JSON enforcing and parsing of the commit messages body, according to the following [JTDSchema](https://ajv.js.org/json-type-definition.html#jtdschematype):

```typescript
interface CommitBodyMetaData {
  job_number?: number
  job_commit?: string
}

interface CommitBodyData {
  namespace: "git-queue.commit-body"
  version: number
  metadata: CommitBodyMetaData
  payload: string
}

const CommitBodySchema: JTDSchemaType<CommitBodyData> = {
  properties: {
    namespace: {enum: ["git-queue.commit-body"]},
    version: {type: 'int32'},
    payload: {type: 'string'},
    metadata: {
      optionalProperties: {
        job_number: {type: 'int32'},
        job_commit: {type: 'string'}
      },
      additionalProperties: true
    }
  },
  additionalProperties: true
}
```

### CODE REVIEW

**Comment by josecelano** [comment]
@yeraydavidrodriguez and I were discussing today if it makes sense to allow the user to add custom metadata.

Pros
- Users could add metadata without defining a new schema for their contents. Right now they could define a JSON schema with custom metadata and the body but it forces them to parse the message body.

Cons
- We have to change the API. We are likely to need an extra argument with an array of custom attributes/metadata.

What do you think @da2ce7?

I've just checked another service (AWS):

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-using-send-message-with-attributes.html

We could have a simpler version like an array of key-value attributes (all of them being strings).

I personally think we should not implement it yet, but only consider that our current schema could change in the future to include this custom attriutes.

**Review by da2ce7 — CHANGES_REQUESTED** [review]
I have reviewed the commit history.  It seems that there has been two merges included in the commit history. Including merging in merge-conflicts and then fixing them in later commit.

I recommend you restage your work on-top of the git-queue master and create new commits.

In this way you can verify that the CI passes for every commit.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -70,7 +70,7 @@ jobs:
           git_repo_dir: ${{ runner.temp }}/temp_git_dir
           queue_name: 'library update - library-aaa'
           action: 'create-job'
-          job_payload: '{"field": "value", "state": "pending"}'
+          job_payload: '{\"field\": \"value\", \"state\": \"pending\"}'
```
[the remark this answers is not in the document]
Comment (yeraydavidrodriguez) [reply .github/workflows/test.yml]: Yes, you are right, Fixed.

**Comment by da2ce7** [comment]
@yeraydavidrodriguez  I have created a squashed commit for the changes that are included in this pull request https://github.com/Nautilus-Cyberneering/git-queue/pull/118

**Review by josecelano — APPROVED** [review]

**Comment by josecelano** [comment]
Moved to a new PR: https://github.com/Nautilus-Cyberneering/git-queue/pull/118

**Pull request closed without merging**